credentials/alts/internal/testutil/testutil.go

```diff
@@ -217,8 +217,13 @@ func (h *FakeHandshaker) processServerStart(req *altspb.StartServerHandshakeReq)
 
 func (h *FakeHandshaker) processNext(req *altspb.NextHandshakeMessageReq, isAssistingClient bool) (*altspb.HandshakerResp, error) {
 	if isAssistingClient {
-		if !bytes.Equal(req.InBytes, []byte("ServerInitServerFinished")) {
-			return nil, fmt.Errorf("unexpected in bytes: got: %v, want: %v", req.InBytes, []byte("ServerInitServerFinished"))
+		// Only check that the one of req.InBytes and
+		// "ServerInitServerFinished" is a prefix-match for the other.
+		// Either case can occur, e.g. because the response from the
+		// server gets split into multiple packets on the wire or the
+		// response also contains encrypted application data.
+		if !bytes.HasPrefix([]byte("ServerInitServerFinished"), req.InBytes) && !bytes.HasPrefix(req.InBytes, []byte("ServerInitServerFinished")) {
```

Comment:
So if `req.InBytes` contains exactly `"S"` then we will send the handshake response?

Should we not wait for the full text to arrive before responding?  (Is this check even worth having if not?)

Reply:
Thanks @dfawley. I initially did not necessarily wait for the full text to arrive (in order to keep the fake handshaker service as simple as possible), but you're correct that this is probably the best way to go since it is a closer representation of the real handshaker service's behavior.

> Can you also take a look at the strange looking test code I pointed out in the issue?
> 
> https://github.com/grpc/grpc-go/blob/417d4b6895679bd9378cb37c2afecf6a292eb267/credentials/alts/alts_test.go#L324-L327

Sure thing. This small code block is working as intended, I think, but there may be better ways to achieve the same thing (I'm not a Go expert so I'm very open to suggestions here). For some context:

* There are 2 global variables, `vmOnGCP` and `once`, in `alts.go` ([ref](https://github.com/grpc/grpc-go/blob/master/credentials/alts/alts.go#L58)). When we run the unit tests in this file and the `newALTS` function is called for the first time, `vmOnGCP` will be set to false.
* In order for `TestFullHandshake` to work, we need to set `vmOnGCP` to true (otherwise, an ALTS handshake will shortcircuit immediately).
* AFAIK, the different tests in `alts_test.go` could run concurrently in different Go routines, so there could be a race condition around which test tries to set `vmOnGCP` first. That's why we've ended up with the current code: if `TestFullHandshake` gets to `once` first, then set to `vmOnGCP` to true. If not, reset it to true anyway.

WDYT? As I said, open to suggestions.